Fix inversion of conditional

weltkante opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Contributes to #4370

## Proposed changes

This regression to Desktop Framework was found while investigating #4370. It is unclear what the effects of this regression are, but [comparing](https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Control.cs,15700) with Desktop Framework it seems its a mistake from refactoring

When running a debug build of WinForms during investigation of #4370 I had turned on `CompModSwitches.ActiveX` and the broken code resulted in warnings of the ActiveX host passing wrong flags (which of course is wrong, using MFC or Office as host you can assume they pass the right flags).

## Customer Impact

- unknown, probably changing behavior of hosted controls due to wrong flags being passed to host

## Regression? 

- yes

## Risk

- low, restoring code to how it was before. There was another recent regression fix for #4370 so this should be fine to accompany the other fix

### Before

- running the sample from #4370 (or generally, hosting a WinForms control in a MFC app or other ActiveX host) outputs warnings when using a WinForms debug build with ActiveX tracing turned on

### After

- no warnings in debug build

## Test methodology

- source inspection
- running the sample from #4370 with a debug build of WinForms, making sure the previously observed warning is gone

## Test environment(s) <!-- Remove any that don't apply -->

- MFC application (stock app from wizard) initializing the WinForms control from the sample code
- Outlook 16.0.13801.20240 (32 bit) hosting the WinForms control from the sample code


###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/4685)

weltkante commented:
@RussKie while investigating #4370 I figured out how to host WinForms controls as ActiveX (which was mostly a problem of MFC being underdocumented and the web starting to forget a lot of docs/examples from the pre- .NET Desktop Framework era). If you think we should add C++/MFC unit tests covering hosting WinForms controls then I could write some code for that. Really depends on how you want to treat ActiveX compatibility going forward.

hughbe commented:
> If you think we should add C++/MFC unit tests covering hosting WinForms controls then I could write some code for that. Really depends on how you want to treat ActiveX compatibility going forward.


Could you share your example? Would love to try testing this out too!

weltkante commented:
@hughbe sure, I've created a cleaned up sandbox project [here](https://github.com/weltkante/NativeWinFormsControlHost) and attached the instructions to create a new project below. Note that you need a custom WinForms build or a recent nightly to get the other (already merged) regression fix from #4370, otherwise the control will not show up.

<details><summary>showing an ActiveX control in an MFC application</summary>

For the managed project:
* use specific bitness to match the native host, e.g. `<PlatformTarget>x86</PlatformTarget>` if the host is building as x86
* add `<EnableComHosting>true</EnableComHosting>` to get the `.comhost.dll` generated
* create a control you want to expose, make it `ComVisible` and give it a `Guid` attribute
* declare an interface and link it via `ComDefaultInterface` to get `IDispatch` support (see [this comment](https://github.com/dotnet/winforms/issues/4370#issuecomment-798828348) for details)
  (this may not be necessary for very basic operation but it is probably a good idea to do anyways)

For the native application:
* Create a new "MFC App" project, mostly with default settings
  * either "Single Document" or "Multiple Documents"
  * switching to project style "MFC Standard" instead of keeping the default since it simplifies the app
  * under advanved features you can disable some more if you want to
    * turning off printing and print preview
    * turning off restart manager support
* in the generated view class add
  ```cpp
  CWnd m_control;
  BOOL Create(LPCTSTR lpszClassName, LPCTSTR lpszWindowName, DWORD dwStyle,
    const RECT& rect, CWnd* pParentWnd, UINT nID, CCreateContext* pContext = NULL) override;
  ```
* in the generated view code add with your controls guid
  ```cpp
  BOOL CMFCApplication1View::Create(LPCTSTR lpszClassName, LPCTSTR lpszWindowName, DWORD dwStyle, const RECT& rect, CWnd* pParentWnd, UINT nID, CCreateContext* pContext)
  {
    CLSID clsid;
    return CView::Create(lpszClassName, lpszWindowName, dwStyle, rect, pParentWnd, nID, pContext)
      && SUCCEEDED(IIDFromString(L"{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}", &clsid))
      && m_control.CreateControl(clsid, NULL, WS_VISIBLE, CRect(0, 0, 200, 200), this, 0);
  }
  ```
  Using `IIDFromString` since I only want to parse a GUID and `CLSIDFromString` does various other auto translations

* add an `app.manifest` with the controls guid to allow running the app without registering the `.comhost.dll` globally
  ```xml
  <assembly manifestVersion="1.0" xmlns="urn:schemas-microsoft-com:asm.v1">
    <assemblyIdentity version="1.0.0.0" name="MyApplication.app"/>
    <file name="ClassLibrary1.comhost.dll">
      <comClass clsid="{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}" threadingModel="Both" />
    </file>
  </assembly>
  ```
  Note that nothing else needs to be in this manifest, VC++ will merge it with the autogenerated manifest.

</details>

RussKie commented:
Regression introduced in #2093

RussKie commented:
> If you think we should add C++/MFC unit tests covering hosting WinForms controls then I could write some code for that.

This would be awesome to have!

weltkante commented:
@RussKie sorry, didn't notice you were waiting with the PR for tests. As mentioned above I don't think this regression is testable in an automated fashion and my comment about testing ActiveX was about testing in general, not related to this PR - but if you want to wait with merging until we have a sample scenario to invoke for manual testing I can get to work and add a variant of the sandbox I linked above to the repo (probably a simplified version, there is no reason to use the full blown VS generated template for simple test scenarios).

RussKie commented:
> you want to wait with merging until we have a sample scenario to invoke for manual testing I can get to work and add a variant of the sandbox I linked above to the repo (probably a simplified version, there is no reason to use the full blown VS generated template for simple test scenarios).

Yes, please. A manually run test is just as good.

weltkante commented:
Added the test scenario, but encountered some issues (you should be able to run it from VS in the x64 configuration)
- x86 doesn't build because the .NET Core SDK can't figure out the right comhost dll (could disable x86 builds in the solution if we don't need them?)
- the amd64 configuration uses the x64 build (this is the same as the already existing native test project does, probably makes no sense and was missed when adding amd64 configurations)
  - I have no idea how to configure a VC++ project for amd64 as its not part of the default template
- this project uses a VC++ project file (instead of CMake) which probably isn't installed on the CI build machines
  - I don't know how to configure CMake projects, but I did setup the VC++ project to honor artifacts directory
  - the native projects outputs itself into the bin folder of the managed control, so I don't have to copy the whole folder over into the native host output; it has its own `obj` build folder though
- I have no idea how to enable `CompModSwitches` via configuration, so I had to add `InternalsVisibleTo` (which otherwise wouldn't be necessary)
  - the `CompModSwitches.ActiveX` setting is required to see the debug output of the failing method call if the fix is not applied

weltkante commented on Winforms.sln:
```diff
@@ -812,6 +819,38 @@ Global
+		{B68FD1FC-EA37-40F2-BBBD-77577A0A1815}.Release|x64.Build.0 = Release|x64
+		{B68FD1FC-EA37-40F2-BBBD-77577A0A1815}.Release|x86.ActiveCfg = Release|Win32
+		{B68FD1FC-EA37-40F2-BBBD-77577A0A1815}.Release|x86.Build.0 = Release|Win32
+		{3129970B-A4EB-46AC-B163-18A5557B11BD}.Debug|Any CPU.ActiveCfg = Debug|x64
+		{3129970B-A4EB-46AC-B163-18A5557B11BD}.Debug|Any CPU.Build.0 = Debug|x64
+		{3129970B-A4EB-46AC-B163-18A5557B11BD}.Debug|arm64.ActiveCfg = Debug|x64
+		{3129970B-A4EB-46AC-B163-18A5557B11BD}.Debug|arm64.Build.0 = Debug|x64
```
this is weird but consistent with the other native project you're building

dreddy-work replied:
App.config did not work? winforms removed dependency but the functionality should be there.

weltkante commented on src/System.Windows.Forms.Primitives/src/Properties/AssemblyInfo.cs:
```diff
@@ -24,6 +24,7 @@
 [assembly: InternalsVisibleTo("MauiRichTextBoxTests, PublicKey=00000000000000000400000000000000")]
 [assembly: InternalsVisibleTo("MauiTestsHelper, PublicKey=00000000000000000400000000000000")]
 [assembly: InternalsVisibleTo("MauiTabControlTests, PublicKey=00000000000000000400000000000000")]
+[assembly: InternalsVisibleTo("NativeHost.ManagedControl, PublicKey=00000000000000000400000000000000")]
```
only necessary because I don't know how to set `CompModSwitches` via configuration settings

dreddy-work replied:
App.config did not work? winforms removed dependency but the functionality should be there.

weltkante replied:
I thought it was removed entirely, I'll try if it works

weltkante replied:
Nope, [the instructions](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.traceswitch?view=net-5.0) don't seem to work. I've tried several variations how to name the `.config` file, and I tried it in a standalone managed project instead of a native host, but neither seemed to work. If there are more up to date instructions I'm happy to try again.

dreddy-work replied:
Strange. For this test project, we can keep as-is. I will play around with winforms projects to see if entire config is not working or some sections are not populated.

weltkante commented on src/System.Windows.Forms/tests/IntegrationTests/NativeHost.ManagedControl/ManagedControl.cs:
```diff
@@ -0,0 +1,40 @@
+    {
+        public ManagedControl()
+        {
+            CompModSwitches.ActiveX.Level = System.Diagnostics.TraceLevel.Verbose;
```
it would be better to do this via some configuration, but how?

weltkante commented on src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:
```diff
@@ -13713,7 +13713,7 @@ unsafe HRESULT Ole32.IOleObject.GetMiscStatus(Ole32.DVASPECT dwAspect, Ole32.OLE
-            if ((dwAspect & Ole32.DVASPECT.CONTENT) != 0)
+            if ((dwAspect & Ole32.DVASPECT.CONTENT) == 0)
             {
                 Debug.WriteLineIf(CompModSwitches.ActiveX.TraceInfo, "AxSource:GetMiscStatus.  Status: ERROR, wrong aspect.");
```
note that observing this message without the above fix applied requires debug build running in a debugger

weltkante commented on src/System.Windows.Forms/tests/IntegrationTests/NativeHost/NativeHost.vcxproj:
```diff
@@ -0,0 +1,178 @@
+  <PropertyGroup Label="UserMacros" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
+    <LinkIncremental>true</LinkIncremental>
+    <OutDir>$(ArtifactsBinDir)NativeHost.ManagedControl\$(PlatformTarget)\$(Configuration)\$(TargetFramework)\</OutDir>
+    <IntDir>$(ArtifactsObjDir)$(ProjectName)\$(PlatformTarget)\$(Configuration)\$(TargetFramework)\</IntDir>
```
I'm recognizing artifcats folders. I'm outputting to `NativeHost.ManagedControl` to avoid making a copy of the whole folder.

weltkante commented on src/System.Windows.Forms/tests/IntegrationTests/NativeHost/NativeHost.cpp:
```diff
@@ -0,0 +1,98 @@
+
+	int ExitInstance() override
+	{
+		// a clean exit is currently not possible, we would have to force a garbage collection
+		// to ensure WinForms had a chance to release its COM objects before MFC exits
+		TerminateProcess(GetCurrentProcess(), S_OK);
```
In debug builds MFC cannot cleanly shutdown due to hitting an assert, WinForms still has references to MFC. I assume forcing a GC might help, but thats not easily doable. As a workaround I just terminate the process.

RussKie commented:
Thank you. I was OOF for a week, and I'll try to review this is within fortnight.

dreddy-work commented:
@weltkante , did you get a chance to look at build failures here?

weltkante commented:
@dreddy-work I specifically called these out as open issues [above](https://github.com/dotnet/winforms/pull/4685#issuecomment-833096072), besides other things which don't immediately result in build failures

> - x86 doesn't build because the .NET Core SDK can't figure out the right comhost dll
>
> [...]
> - this project uses a VC++ project file (instead of CMake) which probably isn't installed on the CI build machines
>
> [...]

I don't know how to resolve this myself; maybe this test project just needs to be excluded from CI builds somehow.

dreddy-work commented:
> @dreddy-work I specifically called these out as open issues [above](https://github.com/dotnet/winforms/pull/4685#issuecomment-833096072), besides other things which don't immediately result in build failures

Thanks. I missed part of it. We may need to get help from people who know infra better here. If these tests can be run against X64 successfully, i think we should unblock this PR and limit tests running to X64 arch only until then and open a new issue to track it.

RussKie commented:
@jkoritzinsky any suggestions on how to build VC++ projects under Arcade?
The solution builds fine from under VS, but from cli (i.e. `.\build.cmd`) it fails with the following:
```
.\winforms\src\System.Windows.Forms\tests\IntegrationTests\NativeHost\NativeHost.vcxproj(27,3): error MSB4019: The imported project "C:\Microsoft.Cpp.Default.props" was not found. Confirm that the expression in the Import declaration "\Microsoft.Cpp.Default.props" is correct, and that the file exists on disk.
```

jkoritzinsky commented:
I think you need to build either from within a Dev Command Prompt, or you need to use Arcade’s feature to build with VS’s MSBuild instead of the dotnet CLI

jkoritzinsky commented:
Specifically, look at the info about specifying a VS version in this documentation for Arcade: https://github.com/dotnet/arcade/blob/main/Documentation/ArcadeSdk.md#globaljson

RussKie commented:
@jkoritzinsky I tried modifying global.json as follows and it didn't work...
```diff
diff --git a/global.json b/global.json
index 41e87f72d..ab604bb19 100644
--- a/global.json
+++ b/global.json
@@ -8,6 +8,10 @@
       "dotnet/x86": [
         "$(VSRedistCommonNetCoreSharedFrameworkx6460PackageVersion)"
       ]
+    },
+    "vs": {
+      "version": "16.11",
+      "components": ["Microsoft.VisualStudio.Component.VC.Tools.x86.x64"]
     }
   },
```

```
  Determining projects to restore...
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\NativeHost\NativeHost.vcxproj : error NU1202: Package Microsoft.NETCore.Platforms 6.0.0-preview.6.21314.2 is not compatible with 
native (native,Version=v0.0). Package Microsoft.NETCore.Platforms 6.0.0-preview.6.21314.2 supports: netstandard1.0 (.NETStandard,Version=v1.0) [C:\Development\winforms\Winforms.sln]
C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\NativeHost\NativeHost.vcxproj : error NU1202: Package Microsoft.NETCore.Platforms 6.0.0-preview.6.21314.2 is not compatible with  
native (native,Version=v0.0) / win. Package Microsoft.NETCore.Platforms 6.0.0-preview.6.21314.2 supports: netstandard1.0 (.NETStandard,Version=v1.0) [C:\Development\winforms\Winforms.sln]
C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\NativeHost\NativeHost.vcxproj : error NU1202: Package Microsoft.NETCore.Platforms 6.0.0-preview.6.21314.2 is not compatible with  
native (native,Version=v0.0) / win-x64. Package Microsoft.NETCore.Platforms 6.0.0-preview.6.21314.2 supports: netstandard1.0 (.NETStandard,Version=v1.0) [C:\Development\winforms\Winforms.sln]
C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\NativeHost\NativeHost.vcxproj : error NU1202: Package Microsoft.NETCore.Platforms 6.0.0-preview.6.21314.2 is not compatible with  
native (native,Version=v0.0) / win-x86. Package Microsoft.NETCore.Platforms 6.0.0-preview.6.21314.2 supports: netstandard1.0 (.NETStandard,Version=v1.0) [C:\Development\winforms\Winforms.sln]
  Failed to restore C:\Development\winforms\src\System.Windows.Forms\tests\IntegrationTests\NativeHost\NativeHost.vcxproj (in 395 ms).
  46 of 47 projects are up-to-date for restore.

Build FAILED.
```

Is this what you had in mind?


@weltkante I've been thinking, and there are three ways we can proceed from here. 
1. We can convert the project to CMake, since we already know how to build it. Here's [my lame attempt](https://github.com/RussKie/winforms/commit/12f09addfd138e15965114cc2e6e2d5f5a6104da), it is getting build from cli, but failing with C++ errors (e.g. `error C2664: 'void CFrameWnd::SetTitle(LPCTSTR)': cannot convert argument 1 from 'const wchar_t [21]' to 'LPCTSTR'`).
2. Since this is a manual-only test, we can leave it as is, exclude from all build configurations, and add a readme with explanation of the intent of this project and how to use it.
3. Combine the previous two :)

jkoritzinsky commented:
Yes that’s what I was thinking. I didn’t realize you were also doing a P2P reference from the C++ project to a managed project. IIRC that doesn’t work too well with the VS project system today.

weltkante commented:
>  I didn’t realize you were also doing a P2P reference from the C++ project to a managed project.

I don't think we do, the native project ist standalone, it consumes the managed project as a COM object, but through a manifest and not by referencing it.

> We can convert the project to CMake

Thanks for getting this started, I have no idea how to use CMake but I can try adding a few more flags like enabling MFC.

weltkante commented:
@RussKie see my [comment](https://github.com/RussKie/winforms/commit/12f09addfd138e15965114cc2e6e2d5f5a6104da#r52309117) over on your branch how to get it compiling, it was using the ANSI win32 methods (apparently default for CMake) not the UNICODE win32 methods (default for VS). I'm not yet convinced everything is set up correctly, in particular that everything gets copied into the right places.

Also, when building as CMake you can't seem to debug from within VS so this is definitely a step back, considering you need to run it in the debugger for verifying the regression got fixed.

RussKie commented:
In this case we can opt out for the option 2 - exclude the c++ project from the build configurations, and add a readme. Then call it done :)

weltkante commented:
I've modified the project to be a no-op when built outside VS - this allows to maintain a single solution file. Dropped a simple readme in the project folder, if you had something specific in mind feel free to update or tell me what to change.

RussKie commented:
I tried building locally, and it builds from cli (with some workarounds if Dev17 installed), as well as in VS.
But I'm having troubles running it:
```
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\artifacts\bin\NativeHost.ManagedControl\x64\Debug\net6.0\NativeHost.exe'. Symbols loaded.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\ntdll.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\kernel32.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\KernelBase.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\user32.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\win32u.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\mfc140ud.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\gdi32.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\advapi32.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\gdi32full.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\msvcp_win.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\msvcrt.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\ucrtbase.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\sechost.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\rpcrt4.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\ole32.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\oleaut32.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\combase.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\vcruntime140d.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\vcruntime140_1d.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\shlwapi.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\ucrtbased.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\imm32.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\ucrtbased.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Unloaded 'C:\Windows\System32\ucrtbased.dll'
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\uxtheme.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.844_none_423537bff6b9828c\comctl32.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\msctf.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\kernel.appcore.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\bcryptprimitives.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\clbcatq.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\artifacts\bin\NativeHost.ManagedControl\x64\Debug\net6.0\NativeHost.ManagedControl.comhost.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\wintrust.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\crypt32.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\msasn1.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\host\fxr\6.0.0-preview.7.21319.2\hostfxr.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\hostpolicy.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\coreclr.dll'. Symbol loading disabled by Include/Exclude setting.
Exception thrown at 0x00007FFC992F4B89 (KernelBase.dll) in NativeHost.exe: 0x04242420 (parameters: 0x0000000031415927, 0x00007FFC03AF0000, 0x000000A17153AAC0).
Profiler was prevented from loading notification profiler due to app settings.  Process ID (decimal): 744.  Message ID: [0x2509].'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\System.Private.CoreLib.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\clrjit.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\icu.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Windows\System32\bcrypt.dll'. Symbol loading disabled by Include/Exclude setting.
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\artifacts\bin\NativeHost.ManagedControl\x64\Debug\net6.0\NativeHost.ManagedControl.dll'. 
'NativeHost.exe' (Win32): Unloaded 'C:\Development\winforms\artifacts\bin\NativeHost.ManagedControl\x64\Debug\net6.0\NativeHost.ManagedControl.dll'
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\artifacts\bin\NativeHost.ManagedControl\x64\Debug\net6.0\NativeHost.ManagedControl.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\artifacts\bin\NativeHost.ManagedControl\x64\Debug\net6.0\System.Windows.Forms.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\System.ComponentModel.Primitives.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\System.Runtime.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\artifacts\bin\NativeHost.ManagedControl\x64\Debug\net6.0\System.Windows.Forms.Primitives.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\System.Runtime.InteropServices.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\System.Drawing.Primitives.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\System.Collections.Specialized.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\System.ComponentModel.TypeConverter.dll'. 
'NativeHost.exe' (Win32): Loaded 'C:\Development\winforms\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.7.21319.2\System.Diagnostics.TraceSource.dll'. 
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x000000A171533C40.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x000000A171533C40.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x000000A171533C40.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x000000A171533C40.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x000000A171533C40.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x000000A171533C40.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x000000A171533C40.
Exception thrown at 0x00007FFC992F4B89 (KernelBase.dll) in NativeHost.exe: 0xE0434352 (parameters: 0xFFFFFFFF80070002, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x00007FFC03AF0000).
Exception thrown at 0x00007FFC992F4B89 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFC992F4B89 (KernelBase.dll) in NativeHost.exe: 0xE0434352 (parameters: 0xFFFFFFFF80070002, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x00007FFC03AF0000).
Exception thrown at 0x00007FFC992F4B89 (KernelBase.dll) in NativeHost.exe: 0xE0434352 (parameters: 0xFFFFFFFF80131604, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, 0x00007FFC03AF0000).
d:\agent\_work\36\s\src\vctools\VC7Libs\Ship\ATLMFC\Src\MFC\occsite.cpp(529) : AppMsg - CoCreateInstance of OLE control {54479E5D-EABC-448C-A767-EAFF17BC28C9} failed.
d:\agent\_work\36\s\src\vctools\VC7Libs\Ship\ATLMFC\Src\MFC\occsite.cpp(530) : AppMsg - >>> Result code: 0x80131604
d:\agent\_work\36\s\src\vctools\VC7Libs\Ship\ATLMFC\Src\MFC\occsite.cpp(531) : AppMsg - >>> Is the control is properly registered?
d:\agent\_work\36\s\src\vctools\VC7Libs\Ship\ATLMFC\Src\MFC\wincore.cpp(734) : AppMsg - Warning: Window creation failed: GetLastError returns 0x00000000
d:\agent\_work\36\s\src\vctools\VC7Libs\Ship\ATLMFC\Src\MFC\winfrm.cpp(626) : AppMsg - Warning: failed to create CFrameWnd.
The thread 0xaa4 has exited with code 0 (0x0).
The thread 0x5f00 has exited with code 0 (0x0).
The thread 0x6f3c has exited with code 0 (0x0).
The thread 0x55ec has exited with code 0 (0x0).
The thread 0x492c has exited with code 0 (0x0).
The thread 0x6ab0 has exited with code 0 (0x0).
The thread 0x5144 has exited with code 0 (0x0).
The program '[744] NativeHost.exe' has exited with code 0 (0x0).
```

I need to dig deeper whether it is a regressions in .NET runtime, or something else.

weltkante commented:
I can confirm getting a first chance exception (breaking into VS if enabled)

> Could not load file or assembly 'System.Drawing.Common, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51

weltkante commented:
I think (part of) the problem is that we're referncing `Microsoft.NETCore.App` but `System.Drawing.Common` is in `Microsoft.WindowsDesktop.App` - it probably doesn't even bother looking there. Not sure what changed and when it did happen.

RussKie commented:
> I can confirm getting a first chance exception (breaking into VS if enabled)
> 
> > Could not load file or assembly 'System.Drawing.Common, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51

@weltkante how did you do that? I tried it in the VS against the latest main, and all I'm getting is 
```
Exception thrown at 0x00007FFB4E064ED9 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x0000000A05CF4390.
Exception thrown at 0x00007FFB4E064ED9 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFB4E064ED9 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x0000000A05CF4390.
Exception thrown at 0x00007FFB4E064ED9 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFB4E064ED9 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x0000000A05CF4390.
Exception thrown at 0x00007FFB4E064ED9 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFB4E064ED9 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x0000000A05CF4390.
Exception thrown at 0x00007FFB4E064ED9 in NativeHost.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFB4E064ED9 in NativeHost.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x0000000A05CF4390.
The program '[14964] NativeHost.exe' has exited with code 0 (0x0).
```


@tommcdon are you able to help in any way?

weltkante commented:
>  I tried it in the VS against the latest main, and all I'm getting is

Did you enable breaking on exceptions? These exceptions are caught otherwise (ultimately returning a failure HRESULT somewhere) and will not break into the debugger, unless you explicitly ask it. Might have to disable "just my code" too.

I think the problem is something differs between generating a normal application and COM hosted DLL, System.Drawing.Common is not included in the dependency layout file and not copied to the output directory. I did try copying it manually but its not loaded, I assume because the dependency json file needs to be correct too.

The other projects seem to include System.Drawing.Common even though its not part of their SDK, not sure what the difference is that triggers this behavior.

RussKie commented:
> I think the problem is something differs between generating a normal application and COM hosted DLL, System.Drawing.Common is not included in the dependency layout file and not copied to the output directory. I did try copying it manually but its not loaded, I assume because the dependency json file needs to be correct too.
> 
> The other projects seem to include System.Drawing.Common even though its not part of their SDK, not sure what the difference is that triggers this behavior.

You could try something like this:
```diff
diff --git a/src/System.Windows.Forms/tests/IntegrationTests/NativeHost.ManagedControl/NativeHost.ManagedControl.csproj b/src/System.Windows.Forms/tests/IntegrationTests/NativeHost.ManagedControl/NativeHost.ManagedControl.csproj
index d139f3387..ac150bf15 100644
--- a/src/System.Windows.Forms/tests/IntegrationTests/NativeHost.ManagedControl/NativeHost.ManagedControl.csproj
+++ b/src/System.Windows.Forms/tests/IntegrationTests/NativeHost.ManagedControl/NativeHost.ManagedControl.csproj
@@ -10,8 +10,13 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\..\..\..\System.Drawing\src\System.Drawing.Facade.csproj" />
     <ProjectReference Include="..\..\..\..\System.Windows.Forms.Primitives\src\System.Windows.Forms.Primitives.csproj" />
     <ProjectReference Include="..\..\..\src\System.Windows.Forms.csproj" />
+    <Reference Include="$(BaseOutputPath)..\System.Drawing.Facade\$(Configuration)\$(TargetFramework)\System.Drawing.dll" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="System.Drawing.Common" Version="$(SystemDrawingCommonPackageVersion)" />
   </ItemGroup>
 
 </Project>
```

> Did you enable breaking on exceptions? These exceptions are caught otherwise (ultimately returning a failure HRESULT back to the MFC host) and will not break into the debugger, unless you explicitly ask it. Might have to disable "just my code" too.

Do you mean the top one here?
![image](https://user-images.githubusercontent.com/4403806/125715482-de00f6d2-f8d0-4816-ab0e-af5d43d77e8c.png)

weltkante commented:
> You could try something like this:

Will do.

> Do you mean the top one here?

No, the exception panel where you specify per-exception whether you want to break on first-chance. I just enable it for all managed exceptions to ensure I'm breaking when something is thrown that is later caught. I'll double check later that it still repros and I didn't miss something, but I think thats what I did.

<details><summary>screenshots</summary>

![image](https://user-images.githubusercontent.com/5845814/125774953-cc99e5cc-c7cd-4be0-9075-12973d4c9339.png)

![image](https://user-images.githubusercontent.com/5845814/125774928-194be271-8778-4a93-9751-d51550863e57.png)

Also "just my code" disabled in the options to break into exceptions when the source is not part of the solution

![image](https://user-images.githubusercontent.com/5845814/125775114-206d1e60-b163-4700-bbea-8fc305c636ec.png)

</details>

RussKie commented:
@ericstj could we get a hand resolving the problem?

I'm puzzled why 'System.Drawing.Common, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' can't be resolved. 
Here's [NativeHost.ManagedControl.deps.json.txt](https://github.com/dotnet/winforms/files/6979935/NativeHost.ManagedControl.deps.json.txt)


NativeHost.ManagedControl references System.Windows.Forms.Primitives project, which in turn references System.Drawing.Common:
https://github.com/dotnet/winforms/blob/65e50657278f6a3af5ccb7ac6def8d5baa5dd490/src/System.Windows.Forms.Primitives/src/System.Windows.Forms.Primitives.csproj#L22

I also tried referencing the facade assembly directly:
```xml
  <ItemGroup>
    <Reference Include="$(BaseOutputPath)..\System.Drawing.Facade\$(Configuration)\$(TargetFramework)\System.Drawing.dll" />
  </ItemGroup>
```
...and I tried referencing the assembly directly:
```xml
  <ItemGroup>
    <PackageReference Include="System.Drawing.Common" Version="$(SystemDrawingCommonPackageVersion)" />
  </ItemGroup>
```

...but none appear didn't appear to make any difference. When run, the app fails with 
![image](https://user-images.githubusercontent.com/4403806/129305401-6e10b6ce-fdba-4331-8474-869c0faa13cb.png)


I also tried force loading the assembly by referencing a type from it like so:
![image](https://user-images.githubusercontent.com/4403806/129305509-182910ae-4bf5-4daf-84f6-e72853711490.png)
...but then the app crashes with 
![image](https://user-images.githubusercontent.com/4403806/129305569-2c28d412-576d-4952-9b13-a6da4ce9aec5.png)

RussKie commented:
I tried this PR against the following reeleases, and it worked:
* [v6.0.0-preview.5.21301.6](https://github.com/dotnet/winforms/tree/v6.0.0-preview.5.21301.6)
* [v6.0.0-preview.6.21353.1](https://github.com/dotnet/winforms/tree/v6.0.0-preview.6.21353.1)

![image](https://user-images.githubusercontent.com/4403806/129510703-1398305d-6acb-432b-853d-bdf42126f5e9.png)

It fails with the errors shown above if run against [v6.0.0-preview.7.21378.6](https://github.com/dotnet/winforms/tree/v6.0.0-preview.7.21378.6)

ericstj commented:
@AaronRobinsonMSFT @jkoritzinsky @elinor-fung  can you suggest how @RussKie might enable some logging to help him troubleshoot this?  Perhaps https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/docs/design/features/host-tracing.md might help.  cc @JeremyKuhne 

As far as I can tell this is a native application that is trying to host .NETCore.  How it locates assemblies would be determined by how the runtime is hosted, which I believe is through [COM activation](https://github.com/dotnet/winforms/commit/e0426f1f5698b7b32e43a551e8ca1dbe8384f882#diff-8af554418331f8e7e5ccc92d1e7279ca21a666c8eb081fa3fa4973cfd0db9873R6) here.  Presumably it should use the runtimeconfig and deps file to do this.

AaronRobinsonMSFT commented:
>Perhaps https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/docs/design/features/host-tracing.md might help

Yes, host tracking is where I would start to see how the assemblies are looked up.

>As far as I can tell this is a native application that is trying to host .NETCore. How it locates assemblies would be determined by how the runtime is hosted, which I believe is through COM activation here. Presumably it should use the runtimeconfig and deps file to do this.

Right. COM Activation in .NET Core follows what is defined by the `.runtimeconfig.json` that is emitted with the COM server's `.comhost.dll`. There was a comment about frameworks here and that could indeed be related since it would require the activation of .NET Core to include that framework &ndash; at least that is how I understand this part of hosting to work.

Note that during a COM activation scenario the COM server and all its dependencies are loaded in a new ALC, not the default ALC.

weltkante commented:
I mean logging is nice and all, but the build system doesn't even copy `System.Drawing.Common.dll` into the output path like it does for other projects that aren't setting `<EnableComHosting>true</EnableComHosting>`. Considering the file isn't there (even though `.deps.json` references it) I'm not surprised it doesn't work, the system has no place where it _could_ be found.

WinForms is using SDK `Microsoft.NET.Sdk` and not `Microsoft.WindowsDesktop.App` (where System.Drawing.Common eventually will live), other projects in WinForms do copy `System.Drawing.Common.dll` to their output directory, but when I enable COM hosting that doesn't happen anymore.

Adding the [logfile](https://github.com/dotnet/winforms/files/6994281/winforms.log) anways in case someone can make something of it.

ericstj commented:
> but the build system doesn't even copy System.Drawing.Common.dll into the output path like it does for other projects that aren't setting `<EnableComHosting>true</EnableComHosting>`

That's a problem.  Your host log indicates it targets the `Microsoft.NETCore.App` shared framework (not WindowsDesktop) which doesn't contain System.Drawing.Common.  The only place for the host to get that library is with your component.   If I look in the buildlog for your component it doesn't copy runtime assets for any PackageReference because `CopyLocalLockFileAssemblies` is false.  Perhaps this is the culprit: https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L36

If you share the working project binlog I might be able to spot the difference to see what changed in the SDK.

A workaround would be to set `CopyLocalLockFileAssemblies` to true.

jkoritzinsky commented:
This is an SDK bug.

The line at https://github.com/dotnet/sdk/blob/f2cc77b60cdc2a4e8b393adf7b855b478ad79671/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L33 sets `EnableDynamicHosting` to `true`, which should set `CopyLocalLockFileAssemblies` to `true`. However, the `Microsoft.PackageDependencyResolution.targets` file is imported earlier in the file before this property is set: 

https://github.com/dotnet/sdk/blob/f2cc77b60cdc2a4e8b393adf7b855b478ad79671/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L15

AaronRobinsonMSFT commented:
>This is an SDK bug.

@jkoritzinsky Why does this work in the non-COM scenario? I've looked at the PR and it seems reasonable but the following statement is concerning without some more context.

> but the build system doesn't even copy System.Drawing.Common.dll into the output path like it does for other projects that aren't setting <EnableComHosting>true</EnableComHosting>.

jkoritzinsky commented:
This works in a non-COM scenario because the "CopyLocalLockFileAssemblies" property is set to true for executable projects, which I believe the other case is (I haven't looked closely at this again recently, but that's my recollection from when I first looked at this)

AaronRobinsonMSFT commented:
@jkoritzinsky So this sounds like yet another one of those cases where a project is actually a self-hosting scenario (e.g., COM server) which is much closer to an EXE than a pure managed assembly project. Is that fair? If so, would it make sense to start building that sort of model in the SDK?

jkoritzinsky commented:
The failing case here that hits the SDK issues isn't self-hosted though, it's a managed COM DLL that is hosted by a native C++ exe, which we already built a model for in the SDK. The "sdk bug" is a bug in this support we added back in .NET Core 3.x (the bug was probably introduced after 3.x).

RussKie commented:
Thank you!

With the following change to the managed control project to run the native host again
```diff
diff --git a/src/System.Windows.Forms/tests/IntegrationTests/NativeHost.ManagedControl/NativeHost.ManagedControl.csproj b/src/System.Windows.Forms/tests/IntegrationTests/NativeHost.ManagedControl/NativeHost.ManagedControl.csproj
index d139f3387..8431433ff 100644
--- a/src/System.Windows.Forms/tests/IntegrationTests/NativeHost.ManagedControl/NativeHost.ManagedControl.csproj
+++ b/src/System.Windows.Forms/tests/IntegrationTests/NativeHost.ManagedControl/NativeHost.ManagedControl.csproj
@@ -14,4 +14,8 @@
     <ProjectReference Include="..\..\..\src\System.Windows.Forms.csproj" />
   </ItemGroup>
 
+  <PropertyGroup>
+    <EnableDynamicLoading>true</EnableDynamicLoading>
+  </PropertyGroup>
+
 </Project>
```

![image](https://user-images.githubusercontent.com/4403806/129641341-7dc7f420-7cfe-4fc6-ab3d-13039e9c094e.png)

RussKie commented:
@weltkante can I get you to rebase, and add the following to NativeHost.ManagedControl.csproj, so we can merge it before we branch off for .NET 7.0
```xml
  <!-- Workaround for https://github.com/dotnet/sdk/pull/19764 -->
  <PropertyGroup>
    <EnableDynamicLoading>true</EnableDynamicLoading>
  </PropertyGroup>
```

Also while you're at it, please fix this too:
![image](https://user-images.githubusercontent.com/4403806/129646223-43c291d4-fc8d-4f37-baba-99c144fe0c99.png)

weltkante commented:
> Also while you're at it, please fix this too:

I don't have that "unused namespace" warning, the event handler in that class uses `EventArgs` from `System` namespace without fully qualifying it.

weltkante commented:
Never mind, I forgot about the new global usings. I don't know why VS didn't pick up on it earlier, but it now does.

RussKie commented:
Thank you! Sorry it took so long.
